client/src/nonblocking.rs

```diff
@@ -0,0 +1,243 @@
+use crate::{
+    handle_program_log, handle_system_log, ClientError, EventContext, Execution, Program,
+    ProgramAccountsIterator, RequestBuilder,
+};
+use anchor_lang::{prelude::Pubkey, AccountDeserialize, Discriminator};
+use futures::stream::StreamExt;
+use futures::Future;
+use solana_account_decoder::UiAccountEncoding;
+use solana_client::{
+    nonblocking::{pubsub_client::PubsubClient, rpc_client::RpcClient},
+    pubsub_client::PubsubClientError,
+    rpc_config::{
+        RpcAccountInfoConfig, RpcProgramAccountsConfig, RpcSendTransactionConfig,
+        RpcTransactionLogsConfig, RpcTransactionLogsFilter,
+    },
+    rpc_filter::{Memcmp, RpcFilterType},
+};
+use solana_sdk::{
+    commitment_config::CommitmentConfig, signature::Signature, signer::Signer,
+    transaction::Transaction,
+};
+use std::{collections::BTreeMap, ops::Deref, pin::Pin, sync::Arc};
+use tokio::{
+    spawn,
+    sync::mpsc::{unbounded_channel, UnboundedReceiver},
+    task::JoinHandle,
+};
+
+type UnsubscribeFunc = Box<dyn FnOnce() -> Pin<Box<dyn Future<Output = ()> + Send>> + Send>;
+
+pub struct EventHandler {
+    client: Arc<PubsubClient>,
+    program_id: Pubkey,
+    config: RpcTransactionLogsConfig,
+    events: BTreeMap<
+        &'static str,
+        (
+            JoinHandle<Result<(), ClientError>>,
+            UnboundedReceiver<UnsubscribeFunc>,
+        ),
+    >,
+}
+
+impl EventHandler {
+    pub fn new(client: PubsubClient, program_id: Pubkey, config: RpcTransactionLogsConfig) -> Self {
+        Self {
+            client: Arc::new(client),
+            program_id,
+            config,
+            events: BTreeMap::new(),
+        }
+    }
+    pub fn suscribe<T: anchor_lang::Event + anchor_lang::AnchorDeserialize>(
+        &mut self,
+        name: &'static str,
```

Comment:
Can we generate subscription ids on subscribe and users can use that id to unsubscribe instead of using a static string?

Reply:
It was my first idea but I don't the right way to make a good id